File: 0701/SeoyeonHong/광고 삽입.py
# https://school.programmers.co.kr/learn/courses/30/lessons/72414
from collections import defaultdict

def stringToSeconds(string):
    h, m, s = string.split(':')
    return int(h) * 3600 + int(m) * 60 + int(s)

def secondsToString(seconds):
    h = seconds // 3600
    seconds %= 3600
    m = seconds // 60
    seconds %= 60
    s = seconds
    return '{0:02d}:{1:02d}:{2:02d}'.format(h, m, s)

def solution(play_time, adv_time, logs):
    answer = 0
    playTime = stringToSeconds(play_time) # 동영상 길이
    advTime = stringToSeconds(adv_time) # 광고 길이
    viewers = defaultdict(int) # 시청자 수
    changes = defaultdict(int) # 초별 시청자 수 변화
    
    for log in logs:
        start, end = log.split("-")
        start = stringToSeconds(start)
        end = stringToSeconds(end)
        changes[start] += 1
        changes[end] -= 1


    viewers[0] = changes[0]
    totalViewers = defaultdict(int)
    totalViewers[0] = viewers[0]
    for t in range(1, playTime+1):
        viewers[t] = viewers[t-1] + changes[t] # 1초 전의 시청자 수 + 추가/감소한 시청자 수

    for t in range(1, playTime + 1):
        totalViewers[t] = totalViewers[t-1] + viewers[t] # 0초 ~ t초 까지의 누적 시청 시간

    maxWatchTime = totalViewers[advTime - 1]

    for t in range(1, playTime - advTime+1):
        watchTime = totalViewers[t+advTime-1] - totalViewers[t-1]
        if maxWatchTime < watchTime:
            maxWatchTime = watchTime
            answer = t

    return secondsToString(answer)

File: 0701/SoohyunKim/광고삽입.js
// https://school.programmers.co.kr/learn/courses/30/lessons/72414

function solution(play_time, adv_time, logs) {
    const playSec = toSeconds(play_time);
    const advSec = toSeconds(adv_time);
    const times = new Array(playSec + 1).fill(0);
    
    for (let log of logs) {
        const [startStr, endStr] = log.split("-");
        const start = toSeconds(startStr);
        const end = toSeconds(endStr);
        times[start] += 1;
        times[end] -= 1;
    }
    
    // 시청자 수
    for (let i = 1; i <= playSec; i++) {
        times[i] += times[i - 1];
    }

    // 누적 시청자
    for (let i = 1; i <= playSec; i++) {
        times[i] += times[i - 1];
    }

    let maxTime = times[advSec - 1];
    let maxStart = 0;

    for (let i = advSec; i < playSec; i++) {
        const total = times[i] - times[i - advSec];
        if (total > maxTime) {
            maxTime = total;
            maxStart = i - advSec + 1;
        }
    }

    return toTime(maxStart);
}

function toSeconds(time) {
    const [hh, mm, ss] = time.split(":").map(Number);
    return hh * 3600 + mm * 60 + ss;
}

function toTime(seconds) {
    const hh = String(Math.floor(seconds / 3600)).padStart(2, '0');
    const mm = String(Math.floor((seconds % 3600) / 60)).padStart(2, '0');
    const ss = String(seconds % 60).padStart(2, '0');
    return `${hh}:${mm}:${ss}`;
}  
